nilearn/regions/connectivity.py

```diff
@@ -0,0 +1,392 @@
+    msz = clust_mask_img.shape
+
+    # Convert the 3D mask array into a 1D vector
+    mskdat = np.reshape(
+        np.asarray(
+            clust_mask_img.dataobj).astype("bool"),
+        prod(msz))
+
+    # Determine the 1D coordinates of the non-zero elements of the mask
+    iv = np.nonzero(mskdat)[0]
+    sz = func_img.shape
+
+    # Reshape fmri data to a num_voxels x num_timepoints array
+    func_data = _safe_get_data(func_img).astype(np.float32)
+    func_data = np.reshape(func_data, (prod(sz[:3]), sz[3]))
+
+    # Mask the datset to only the in-mask voxels
+    func_data = func_data[iv, :]
+    func_data_sz = func_data.shape
+
+    # Z-score fmri time courses, this makes calculation of the
+    # correlation coefficient a simple matrix product
+    func_data_s = np.tile(np.std(func_data, 1), (func_data_sz[1], 1)).T
+
+    # Replace 0 with large number to avoid div by zero
+    func_data_s[func_data_s == 0] = 1000000
+    func_data_m = np.tile(np.mean(func_data, 1), (func_data_sz[1], 1)).T
+    func_data = (func_data - func_data_m) / func_data_s
+
+    # Set values with no variance to zero
+    func_data[func_data_s == 0] = 0
+    func_data[np.isnan(func_data)] = 0
+
+    # Remove voxels with zero variance, do this here so that the mapping will
+    # be consistent across subjects
+    vndx = np.nonzero(np.var(func_data, 1) != 0)[0]
+    iv = iv[vndx]
+    m = len(iv)
+    print(m, " # of non-zero valued or non-zero variance voxels in the mask")
+
+    # Construct a sparse matrix from the mask
+    msk = csc_matrix(
+        (vndx + 1, (iv, np.zeros(m))), shape=(prod(msz), 1), dtype=np.float32
+    )
+
+    sparse_i = []
+    sparse_j = []
+    sparse_w = [[]]
+
+    for i in range(0, m):
+        if i % 1000 == 0:
+            print("voxel #", i)
+
+        # Convert index into 3D and calculate neighbors, then convert resulting
+        # 3D indices into 1D
+        ndx1d = indx_3dto1d(indx_1dto3d(iv[i], sz[:-1]) + neighbors, sz[:-1])
+
+        # Convert 1D indices into masked versions
+        ondx1d = msk[ndx1d].todense()
+
+        # Exclude indices not in the mask
+        ndx1d = ndx1d[np.nonzero(ondx1d)[0]].flatten()
+        ondx1d = np.array(ondx1d[np.nonzero(ondx1d)[0]])
+        ondx1d = ondx1d.flatten() - 1
+
+        # Keep track of the index corresponding to the "seed"
+        nndx = np.nonzero(ndx1d == iv[i])[0]
+
+        # Extract the time courses corresponding to the "seed" and 3D
+        # neighborhood voxels
+        tc = np.array(func_data[ondx1d.astype("int"), :])
+
+        # Ensure that the "seed" has variance, if not just skip it
+        if np.var(tc[nndx, :]) == 0:
+            continue
+
+        # Calculate functional connectivity maps for "seed" and 3D neighborhood
+        # voxels
+        R = np.corrcoef(np.dot(tc, func_data.T) / (sz[3] - 1))
+
+        if np.linalg.matrix_rank(R) == 1:
+            R = np.reshape(R, (1, 1))
+
+        # Set nans to 0
+        R[np.isnan(R)] = 0
```

Comment:
thresholding should be done at the array level.

Reply:
Do you mean out = (R < thresh) * R  or out = np.where(R < thresh, R, 0)? The former is faster I believe